pkg/scheduler/predicates.go

```diff
@@ -169,6 +169,119 @@ func PodFitsHost(pod api.Pod, existingPods []api.Pod, node string) (bool, error)
 	return pod.Spec.Host == node, nil
 }
 
+type NodeLabelChecker struct {
+	info     NodeInfo
+	labels   []string
+	presence bool
+}
+
+func NewNodeLabelPredicate(info NodeInfo, labels []string, presence bool) FitPredicate {
+	labelChecker := &NodeLabelChecker{
+		info:     info,
+		labels:   labels,
+		presence: presence,
+	}
+	return labelChecker.CheckNodeLabelPresence
+}
+
+// CheckNodeLabelPresence checks whether a particular label exists on a minion or not, regardless of its value
+// Consider the cases where the minions are placed in regions/zones/racks and these are identified by labels
+// In some cases, it is required that only minions that are part of ANY of the defined regions/zones/racks be selected
+//
+// Alternately, eliminating minions that have a certain label, regardless of value, is also useful
+// A minion may have a label with "retiring" as key and the date as the value
+// and it may be desirable to avoid scheduling new pods on this minion
+func (n *NodeLabelChecker) CheckNodeLabelPresence(pod api.Pod, existingPods []api.Pod, node string) (bool, error) {
+	var exists bool
+	minion, err := n.info.GetNodeInfo(node)
+	if err != nil {
+		return false, err
+	}
+	for _, label := range n.labels {
+		exists = labels.Set(minion.Labels).Has(label)
+		if (exists && !n.presence) || (!exists && n.presence) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+type ServiceAffinity struct {
+	podLister     PodLister
+	serviceLister ServiceLister
+	nodeInfo      NodeInfo
+	labels        []string
+}
+
+func NewServiceAffinityPredicate(podLister PodLister, serviceLister ServiceLister, nodeInfo NodeInfo, labels []string) FitPredicate {
```

Comment:
I got a bit confused what you're trying to do here. I guess what you're trying to do is add an implicit selector requiring some particular value V for label L to a pod which does not have any selector for L, if L is listed in the ServiceAffinity object that is passed into the function and some other pod from the same service is already scheduled onto a minion that has value V for label L? If so, please use that as the comment. If not, please explain more clearly.

As for whether this is useful as a predicate rather than as a priority function, it's hard to say. I suspect a priority function is more useful, since it allows you to say "spread across as few zones as possible" and thus can handle the case where you have more pods than you have resources in a single zone. However, I can also see the usefulness of what you implemented here, so I would leave it for now.